lib/rouge/lexers/mason.rb

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*- #
+# frozen_string_literal: true
+
+module Rouge
+  module Lexers
+    class Mason < TemplateLexer
+      title 'Mason'
+      desc 'HTML::Mason'
+      tag 'mason'
+      filenames '*.mi', '*.mc', '*.mas', '*.m', '*.mhtml', '*.mcomp', 'autohandler', 'dhandler'
+      mimetypes 'text/x-mason', 'application/x-mason'
+
+      def initialize(*)
+        super
+        @perl = Perl.new
+      end
+  
+      def self.detect?(text)
+        return false if text.doctype?(/[html|xml]/)
+        return true if text.doctype?
+      end
+
+      textblocks = %w(text doc)
+      perlblocks = %w(args flags attr init once shared perl cleanup filter)
+      components = %w(def method)
+  
+      state :root do
+        mixin :mason_tags
+      end
+
+      state :mason_tags do
+        rule /\s+/, Text::Whitespace
+
+        rule /<%(#{textblocks.join('|')})>/i, Keyword::Constant, :text_block
+
+        rule /<%(#{perlblocks.join('|')})>/i, Keyword::Constant, :perl_block
+
+        rule /(<%(#{components.join('|')}))([^>]*)(>)/i do |m|
+          token Keyword::Constant, m[1]
+          token Name, m[3]
+          token Keyword::Constant, m[4]
+          push :component_block
+        end
+
+        # other perl blocks
+        rule /<%([a-zA-Z_]*)>/i, Keyword::Constant, :other_perl_blocks
+        
+        # perl comment
+        rule /^(#.*)$/, Comment
+
+        # perl line
+        rule /^%(.*)$/ do |m|
+          token Keyword::Constant, '%'

```

Comment:
I think it would be 'safer' to capture the percent sign in the regex and then refer to it here. Note that that will mean the next line should refer to `m[2]`.